core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,198 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+class UpdateSnapshotReferencesOperation implements PendingUpdate<Map<String, SnapshotRef>> {
+
+  private final TableOperations ops;
+  private final Map<String, SnapshotRef> updatedRefs;
+  private TableMetadata base;
+
+  UpdateSnapshotReferencesOperation(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.updatedRefs = Maps.newHashMap(base.refs());
+  }
+
+  @Override
+  public Map<String, SnapshotRef> apply() {
+    return refDiff(base.refs(), updatedRefs);
+  }
+
+  @Override
+  public void commit() {
+    TableMetadata updated = internalApply();
+    ops.commit(base, updated);
+  }
+
+  public UpdateSnapshotReferencesOperation createBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef branch = SnapshotRef.branchBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, branch);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation createTag(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef tag = SnapshotRef.tagBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, tag);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeBranch(String name) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such branch with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.remove(name);
```

Comment:
This will also return the ref that was removed. It would be better to call `Snapshot ref = updatedRefs.remove(name)` and then run the validations on the returned ref.

If validation fails, the operation is no longer valid, so don't worry about changing state and then failing.